### PR TITLE
Make ZStream#aggregate safe to pull again

### PR DESCRIPTION
**Pull request by vasilmkd**
I am extremely grateful for how many tests we have covering all subtle edge cases of this combinator.

@iravid

### CODE REVIEW

**Comment by vasilmkd** [comment]
I'll fix 2.11 compilation issues but can anyone take a look what the dotty issue is? @adamgfraser

**Comment by vasilmkd** [comment]
@iravid you have written a lot of local classes and path dependent types before, can you please take a look at the compilation issues? I'm running out of ideas.

**Comment by regiskuckaertz** [comment]
What if you remove the `final` modifier?

**Comment by vasilmkd** [comment]
> What if you remove the `final` modifier?

Tried, sadly no. :cry:

**Comment by regiskuckaertz** [comment]
Hmm, another idea would be to hoist the definition at the top of the method, rather than there inside the pattern match 🤔

**Comment by vasilmkd** [comment]
Yes, I thought of that too, but I need the matched sink value's `State` type.

**Comment by vasilmkd** [comment]
Dotty seems to really dislike local definitions. The error had something to do with the significant whitespace syntax, which I thought we had solved before.

Anyways, it's finally ready for review. @iravid

**Comment by adamgfraser** [comment]
@vasilmkd I don't think it is related to the significant whitespace proposal, since we have disabled that. I believe it is related to defining a local type that shadows an abstract type in an enclosing scope (`sealed trait State` in `aggregate` and `type State` in `ZSink`). You should be able to avoid it by just renaming your ADT to something like `AggregateState` and then creating a type alias `type State = AggregateState` as done [here](https://github.com/zio/zio/pull/1928/files#diff-7e76fb4cea322437475668ac73762d6aR729).

**Comment by vasilmkd** [comment]
@adamgfraser Thanks, I'll try that. The thing is, I think `ZStream.scala` compiles correctly without these workarounds.

File: streams/shared/src/main/scala/zio/stream/StreamEffect.scala [hunk 1]
```diff
@@ -232,58 +232,73 @@ private[stream] class StreamEffect[-R, +E, +A](val processEffect: ZManaged[R, No
...
-                      val (newState, newLeftovers) = sink.stepChunkPure(state, leftovers)
-                      leftovers = newLeftovers
-                      go(newState, true)
-                    } else {
-                      val a = thunk()
-                      go(sink.stepPure(state, a), true)
-                    }
+              var state: AggregateState[sink.State, A1] = AggregateState.Pull(sink.initialPure, false)
+
+              @annotation.tailrec
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/zio/stream/StreamEffect.scala]: 🚀 ✨

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -586,70 +586,92 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
...
-                              leftoversRef
-                                .update(_ ++ leftovers)
-                                .when(leftovers.nonEmpty)
-                                .as(b)
-                          }
-                    },
+    ZStream
+      .managed(managedSink)
+      .flatMap { sink =>
+        sealed abstract class State
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: Looks like you could reuse `AggregateState` here?

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: Sure, why not.

File: streams/shared/src/main/scala/zio/stream/StreamEffect.scala (hunk 1 above)
Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/StreamEffect.scala]: 🤣 🤣 🤣

**Comment by vasilmkd** [comment]
I tried to minimize code duplication as much as possible. It took quite some time to write though. :)
